SourceX/controls/plrctrls.cpp

```diff
@@ -0,0 +1,693 @@
+	return { -1, -1 };
+}
+
+void checkItemsNearby(bool interact)
+{
+	for (int i = 0; i < MAXITEMS; i++) {
+		if (checkNearbyObjs(item[i]._ix, item[i]._iy, 1).x != -1 && item[i]._iSelFlag > 0 && item[i]._itype > -1) {
+			pcursitem = i;
+			if (dItem[item[i]._ix][item[i]._iy] <= 0)
+				continue;
+			if (interact) {
+				//sprintf(tempstr, "FOUND NEARBY ITEM AT X:%i Y:%i SEL:%i", item[i]._ix, item[i]._iy, item[i]._iSelFlag);
+				//NetSendCmdString(1 << myplr, tempstr);
+				LeftMouseCmd(false);
+			}
+			return; // item nearby, don't find objects
+		}
+	}
+	if (newCurHidden)
+		pcursitem = -1;
+	//sprintf(tempstr, "SCANNING FOR OBJECTS");
+	//NetSendCmdString(1 << myplr, tempstr);
+	for (int i = 0; i < MAXOBJECTS; i++) {
+		if (checkNearbyObjs(object[i]._ox, object[i]._oy, 1).x != -1 && object[i]._oSelFlag > 0 && object[i]._otype > -1 && currlevel) { // make sure we're in the dungeon to scan for objs
+			pcursobj = i;
+			if (interact) {
+				LeftMouseCmd(false);
+			}
+			return;
+		}
+	}
+	if (newCurHidden)
+		pcursobj = -1;
+}
+
+void checkTownersNearby(bool interact)
+{
+	if (pcursitem != -1)
+		// Items take priority over towners because the player can move
+		// items but not towners.
+		return;
+	for (int i = 0; i < 16; i++) {
+		if (checkNearbyObjs(towner[i]._tx, towner[i]._ty, 2).x != -1) {
+			if (towner[i]._ttype == -1)
+				continue;
+			pcursmonst = i;
+			if (interact) {
+				TalkToTowner(myplr, i);
+			}
+			break;
+		}
+	}
+}
+
+bool checkMonstersNearby(bool attack)
+{
+	int closest = 0;                 // monster ID who is closest
+	coords objDistLast = { 99, 99 }; // previous obj distance
+	// The first MAX_PLRS monsters are reserved for players' golems.
+	for (int i = MAX_PLRS; i < MAXMONSTERS; i++) {
+		int d_monster = dMonster[monster[i]._mx][monster[i]._my];
+		if (monster[i]._mFlags & MFLAG_HIDDEN || monster[i]._mhitpoints <= 0) // monster is hiding or dead, skip
+			continue;
+		if (d_monster && dFlags[monster[i]._mx][monster[i]._my] & BFLAG_LIT) {                                                                          // is monster visible
+			if (monster[i].MData->mSelFlag & 1 || monster[i].MData->mSelFlag & 2 || monster[i].MData->mSelFlag & 3 || monster[i].MData->mSelFlag & 4) { // is monster selectable
+				coords objDist = checkNearbyObjs(monster[i]._mx, monster[i]._my, 6);
+				if (objDist.x > -1 && objDist.x <= objDistLast.x && objDist.y <= objDistLast.y) {
+					closest = i;
+					objDistLast = objDist;
+				}
+			}
+		}
+	}
+	if (closest > 0) { // did we find a monster
+		pcursmonst = closest;
+		//sprintf(tempstr, "NEARBY MONSTER WITH HP:%i", monster[closest]._mhitpoints);
+		//NetSendCmdString(1 << myplr, tempstr);
+	} else if (closest > 0) { // found monster, but we don't want to attack it
+		return true;
+	} else {
+		pcursmonst = -1;
+		return false;
+	}
+	if (attack) {
+		static DWORD attacktick = 0;
+		DWORD ticks = GetTickCount();
+		if (ticks - attacktick > 100) { // prevent accidental double attacks

```

Comment:
logic frames are 50ms, maybe that a better value, an maybe we should keep track of them and simply skip the whole function based on that.